webapp/graphite/render/datalib.py

```diff
@@ -217,3 +221,102 @@ def nonempty(series):
       return True
 
   return False
+
+
+def _extend_for_new_metrics(seriesList, pathExpr, startTime, endTime):
+  ################################################################################
+  ################################################################################
+  # Search for Carbon-cache in case for some metrics that has not been flushed yet
+
+  # If the series.name has alread existed in seriesList, then do nothing
+  # else extend the data in cache to seriesList.
+
+  # Only do this if step == lowest
+  # 1. Get infos from archive
+  #    based on time range to determine step
+  # 2. Only query cache if step == min(archive['secondsPerPoints'])
+  # 3. preprocess the pathExpr to be real metric path, then use
+  #    carbonlink to query MemCache
+  # 4. Edge case: pathExpr has wildcard (Ignored it for now)
+  #    For now, let's just support specific individual metric query
+
+  # Add a function like
+  # extend_for_new_metrics(seriesList, pathExpr)
+
+  clean_patterns = pathExpr.replace('\\', '')
+  has_wildcard = clean_patterns.find('{') > -1 or clean_patterns.find('[') > -1 or clean_patterns.find('*') > -1 or clean_patterns.find('?') > -1
+
+  # 1) CarbonLink has some hosts
+  # 2) has no wildcard
+  if CarbonLink.hosts and not has_wildcard:
```

Comment:
I see a lot of code here that already exists in other modules. can we factor this could out to a separate module and import them?

duplicated code is very hard to maintain: usually people modify in one place but not the other.